doc/encoding-notes.txt

```diff
@@ -24,3 +24,22 @@ URL paths:
 
     On disk:
         use internal format
+
+
+Hostname domain:
+    Input - From HTML:
+        handle with idna (ASCII)
+        Example:
+            http://čau-☺フ.даль.рф:8000/
+            b'xn--http://au-..:8000/-0gc871nnc3hue4drjv997fp38b'
```

Comment:
Only the hostname should be encoded, not the protocol (`http://`) or the port/path.

Reply:
Maybe Werkzeug's [url_parse](https://werkzeug.palletsprojects.com/en/1.0.x/urls/?highlight=url#werkzeug.urls.url_parse), which returns a result with has [ascii_host](https://werkzeug.palletsprojects.com/en/1.0.x/urls/?highlight=url#werkzeug.urls.BaseURL.ascii_host), might be helpful?

Reply:
We can discuss these conversions:

```shell
>>> p = urls.url_parse("http://čau-☺フ.даль.рф:8000/")
>>> p = p.ascii_host
>>> p2 = idna.ToASCII("čau-☺フ.даль.рф:8000")
>>> p2 == p
False
>>> p
'xn--au--eqa2078bpkn.xn--80ahw2e.xn--p1ai'
>>> p2
b'xn--au-..:8000-ljb457gqb3fqd2c5g9019d4dub'
>>> p2.decode('punycode')
'xn--a⚨uç-͎ᷥ͊..:8͑0͗00́̈́'
>>> p2.decode()
'xn--au-..:8000-ljb457gqb3fqd2c5g9019d4dub'
>>> str(p2)
"b'xn--au-..:8000-ljb457gqb3fqd2c5g9019d4dub'"
>>> bytes(p)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: string argument without an encoding
```

Reply:
Flask returns this encoding:

http://xn--au-:8000-jbb7826gknya/second_page.html

```shell
>>> url_parse('http://xn--au-:8000-jbb7826gknya/second_page.html').ascii_host
'xn--au-'
```

Reply:
Only the hostname should be encoded (not the port).
Also, each part of the hostname should be encoded separately, so that e.g. `čeština.example.com` still ends with `.example.com` when encoded.